src/utils.zig

```diff
@@ -1907,6 +1907,125 @@ test "read/write mixed" {
+			self.dense.deinit(allocator);
+			self.sparse.deinit(allocator);
+			self.freeList.deinit(allocator);
+		}
+
+		pub fn contains(self: *Self, id: idType) bool {
+			return id < self.sparse.items.len and self.sparse.items[id] != noValue;
+		}
+
+		pub fn add(self: *Self, allocator: NeverFailingAllocator, value: T) idType {
+			const sparseId: idType = self.freeList.popOrNull() orelse @intCast(self.sparse.items.len);
+
+			if(sparseId == self.sparse.items.len) {
+				self.sparse.append(allocator, noValue);
+			}
+
+			const denseId: idType = @intCast(self.dense.items.len);
+			self.sparse.items[sparseId] = denseId;
+			self.dense.append(allocator, .{.value = value, .id = sparseId});
+
+			return sparseId;
+		}
+
+		pub fn remove(self: *Self, allocator: NeverFailingAllocator, sparseId: idType) void {
+			if(!self.contains(sparseId)) return;
+
+			const denseId = self.sparse.items[sparseId];
+			if(denseId == noValue) return;
+
+			self.freeList.append(allocator, sparseId);
+
+			if(self.dense.items.len == 1) {
+				_ = self.dense.pop();
+			} else {
+				self.dense.items[denseId] = self.dense.pop();
+			}
+
+			self.sparse.items[sparseId] = noValue;
+			self.sparse.items[self.dense.items[denseId].id] = denseId;
+		}
+
+		pub fn get(self: *Self, id: idType) GetError!*T {
+			if(id >= self.sparse.items.len) return GetError.idOutOfBounds;
+			const index = self.sparse.items[id];
+			if(index == noValue) return GetError.valueDoesntExist;
+			return &self.dense.items[index].value;
+		}
+	};
+}
+
+const SparseSetTest = struct {
+	var testingAllocator = main.heap.ErrorHandlingAllocator.init(std.testing.allocator);
+	var allocator = testingAllocator.allocator();
+};
+
+test "SparseSet/adding" {
+	var set: SparseSet(u32, u32) = .{};
+	defer set.deinit(SparseSetTest.allocator);
+
+	const id = set.add(SparseSetTest.allocator, 5);
+	try std.testing.expectEqual(id, 0);
+	try std.testing.expectEqual((try set.get(id)).*, 5);
+}
+
+test "SparseSet/removing" {
+	var set: SparseSet(u32, u32) = .{};
+	defer set.deinit(SparseSetTest.allocator);
+
+	const expectSecond: u32 = 100;
+
+	const firstId = set.add(SparseSetTest.allocator, 5);
+	const secondId = set.add(SparseSetTest.allocator, expectSecond);
+
+	set.remove(SparseSetTest.allocator, firstId);
+
+	try std.testing.expectEqual((try set.get(secondId)).*, expectSecond);
+}
+
+test "SparseSet/reusing" {
+	var set: SparseSet(u32, u32) = .{};
+	defer set.deinit(SparseSetTest.allocator);
+
+	const expectSecond = 100;
+	const expectNew = 10;
+
+	const firstId = set.add(SparseSetTest.allocator, 5);
+	const secondId = set.add(SparseSetTest.allocator, expectSecond);
+
+	set.remove(SparseSetTest.allocator, firstId);
+
+	const newId = set.add(SparseSetTest.allocator, expectNew);
+
+	try std.testing.expectEqual(firstId, newId);
+
+	try std.testing.expectEqual((try set.get(secondId)).*, expectSecond);
+	try std.testing.expectEqual((try set.get(newId)).*, expectNew);
+}
```

Comment:
All the tests are doing almost the same thing, so I'd suggest to merge them.

Reply:
I don't agree, tests should be simple and should test specific edge cases. We added few more test cases testing few more edge cases. Test should follow arrange, act, assert (AAA) rule, which we do here, otherwise they get complicated and their failure is ambiguous. If we were to merge the tests, we would either neglect verifying some of the outputs, or we would have multiple of act-assert segments.

Reply:
Well I was just saying this because before the last test was basically doing the same thing as the two previous tests, but the new tests are a lot better in that regard.